### PR TITLE
Refactor(build): TypeScript-ize webpack config files

### LINKED ISSUE #3082: Question: TypeScript-ize webpack-related files?
Webpack [supports configuration files in TypeScript](https://webpack.js.org/configuration/configuration-languages/#typescript), but needs one more `tsconfig-for-webpack.json` and some special environment variables. So I'm not sure if it is worth to TypeScript-ize those files as well.

### PR DESCRIPTION
**Pull request by MaikoTan**
Close #3082

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,18 +31,24 @@
...
+    "@types/webpack": "^5.28.0",
+    "@types/webpack-dev-server": "^3.11.4",
     "@typescript-eslint/eslint-plugin": "^4.15.0",
     "@typescript-eslint/parser": "^4.15.1",
     "@zoltu/typescript-transformer-append-js-extension": "^1.0.1",
     "babel-loader": "^8.2.2",
     "chai": "^4.2.0",
     "chalk": "^2.4.2",
-    "clean-webpack-plugin": "^3.0.0",
+    "clean-webpack-plugin": "^4.0.0-alpha.0",
```
Comment (quisquous) [inline package.json:49]: Did you need something from this alpha version?

File: tsconfig-for-webpack.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "compilerOptions": {
```
Comment (quisquous) [inline tsconfig-for-webpack.json:2]: What do you think about:
```json
    "strict": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": false,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
```

File: package.json (hunk 1 above)
Comment (MaikoTan) [reply package.json:49]: The exported class of v3.x is not compatible with the plugin interface of webpack 5, I didn't investigate why but update to 4.0.0 fixed this.

**Comment by MaikoTan** [comment]
There's still issuses in building, let me figure it out.

**Comment by quisquous** [comment]
```shell
$ npm start

> cactbot@ start C:\Users\enne\cactbot
> cross-env TS_NODE_PROJECT="tsconfig-for-webpack.json" webpack-dev-server --con
fig webpack/webpack.dev.ts

C:\Users\enne\cactbot\webpack\webpack.dev.ts:1
import { mergeWithRules } from 'webpack-merge';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:979:16)
    at Module._compile (internal/modules/cjs/loader.js:1027:27)
```

Is `module: commonjs` wrong? Or is there something wrong with my setup?

**Comment by MaikoTan** [comment]
> ```shell
> $ npm start
> 
> > cactbot@ start C:\Users\enne\cactbot
> > cross-env TS_NODE_PROJECT="tsconfig-for-webpack.json" webpack-dev-server --con
> fig webpack/webpack.dev.ts
> 
> C:\Users\enne\cactbot\webpack\webpack.dev.ts:1
> import { mergeWithRules } from 'webpack-merge';
> ^^^^^^
> 
> SyntaxError: Cannot use import statement outside a module
>     at wrapSafe (internal/modules/cjs/loader.js:979:16)
>     at Module._compile (internal/modules/cjs/loader.js:1027:27)
> ```
> 
> Is `module: commonjs` wrong? Or is there something wrong with my setup?

Yeah, I met this issue too.
And I remove those lines you suggested in `tsconfig-for-webpack.json` then it work normally.

**Pull request merged**